8360934: Add AVX-512 intrinsics for ML-KEM - enhancement on AVX512_VBMI

smemery opened the pull request:
This change allows use of the AVX512_VBMI instruction set to further optimize decompression/parsing of polynomial coefficients for ML-KEM.  The speedup gained in the ML-KEM benchmarks for key generation is between 0.4 to 0.5%, encapsulation is  0.2 to 1.7%, and decapsulation is 0.3 to 2.0%.

Thank you to @sviswa7 and @ferakocz for their help in working through the early stages of this code with me.

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8360934](https://bugs.openjdk.org/browse/JDK-8360934): Add AVX-512 intrinsics for ML-KEM - enhancement on AVX512_VBMI (**Enhancement** - P5)


### Reviewers
 * [Volodymyr Paprotski](https://openjdk.org/census#vpaprotski) (@vpaprotsk - Committer) Review applies to [373b1339](https://git.openjdk.org/jdk/pull/28815/files/373b133958b8764e79be8628ac773215e43119a5)
 * [Jatin Bhateja](https://openjdk.org/census#jbhateja) (@jatin-bhateja - **Reviewer**)

### Contributors
 * Sandhya Viswanathan `<sviswanathan@openjdk.org>`

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/28815/head:pull/28815` \
`$ git checkout pull/28815`

Update a local copy of the PR: \
`$ git checkout pull/28815` \
`$ git pull https://git.openjdk.org/jdk.git pull/28815/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 28815`

View PR using the GUI difftool: \
`$ git pr show -t 28815`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/28815.diff">https://git.openjdk.org/jdk/pull/28815.diff</a>

</details>
<details><summary>Using Webrev</summary>

[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/28815#issuecomment-3674017799)
</details>

smemery committed:
8360934: Add AVX-512 intrinsics for ML-KEM - enhancement on AVX512_VBMI and AVX512_VBMI2

smemery committed:
Fix whitespaces

smemery committed:
Switch from vpshldvw to vpsrlvw

smemery committed:
Reverse ordering for vpermb and vpsrlvw instructions

smemery committed:
Remove wrong mask from evpsrlvw

smemery committed:
Swap parameter operation with source

smemery committed:
Merge with mainline

smemery commented:
/label add security

smemery committed:
Update copyright year

vpaprotsk commented on src/hotspot/cpu/x86/stubGenerator_x86_64_kyber.cpp:
```diff
@@ -822,10 +855,64 @@ address generate_kyber12To16_avx512(StubGenerator *stubgen,
+
+      __ addptr(condensed, 192);
+      __ addptr(parsed, 256);
+      __ subl(parsedLength, 128);
```
(128 instead of 256 here because `parsedLength` is an index to an `short` array..)

I am confused by the stride. The `twelve2Sixteen()` seems to (almost) guarantee that the parsed length is a multiple of 64 (last block can be 48 bytes). This would imply a stride of 128 bytes for `parsed`. And 96 for `condensed`.

This is exactly how the existing code already behaves so I am less concerned, but I would like an explanation why it works?

smemery replied:
I believe the numbers are right: with each pass 256 bytes of coefficients are `parsed` into the parse buffer.  This means that half of the coefficients have been processed (`parseLength` = 128).  Would having a comment stating as such be sufficient for your concerns?

vpaprotsk replied:
I wasn't as clear in my question. The asm is indeed processing the bytes in the increment. What I was trying to convince myself about.. 'how come we are not reading past the end of the array. Or are we?'.

On one hand, this is exactly what the existing asm code does, so I will assume that its correct. However, on the java side/version of this code, I could only convince myself about processing ~two AVX512 vectors at a time, not four.

So either I cant count, or there is some further (implicit) restrictions on the callers of `twelve2Sixteen`

vpaprotsk replied:
Yep, thats exactly the assert I was looking at as well.. looks to me like its dividing the 'expanded-short-array-length' by 64 and ensuring the remainder is zero (ignoring the 48 for a bit.. and the condensed-length check).

(for simplicity) So the 'expanded' array length should be a multiple of 64; i.e. 128-bytes. But we stride the expanded array by 256 bytes? (and parsedLength by 128-shorts..)

I haven't checked the callers of `twelve2Sixteen` but I suspect that the length of the expanded array is always a multiple of 256-bytes (128-shorts).. in which case, the assert is 'incomplete'?

smemery replied:
I've filed bug:

[JDK-8374755](https://bugs.openjdk.org/browse/JDK-8374755) ML-KEM's 12-bit decompression uses incorrect assertions

to track this issue.

vpaprotsk reviewed (commented):
"Insert 0b0000 nibble after every third nibble". I only have two questions, looks good otherwise.


PS: things I've considered:

- Loop controls?
  - ML_KEM.java guarantees  (per callee comment and assert) lengths are multiple of 64
  - also same as original code
- Why not simply a vpermb? Have zeroes already from the masked load with k1..
  - shuffle granularity is actually 4-bits, not 8-bits
- logical shift already zeroes top bits, so `vpand` not required?
  - odd columns not shifted, so still have extra bits that need clearing
- Why VBMI?
  - needed for `evpermb`

smemery committed:
8360934: Add AVX-512 intrinsics for ML-KEM - enhancement on AVX512_VBMI
Change Swap to Dup named function/variable
Check for only VBMI support (not VBMI2)

vpaprotsk reviewed (approved):
Looks good to me

smemery committed:
Merge with mainline

vpaprotsk reviewed (approved):
> @vpaprotsk I've reran related regression tests and benchmarks after implementing your code review comments and remerging with the master branch. These have all came back with the expected results. Could you reapprove after the merge commit? Thank you.

Sure! Though my approval wont let you integrate, I am a committer, not a reviewer. Maybe we can ask @ascarpino

smemery committed:
Update to use OptoLoopAlignment for VBMILoop

jatin-bhateja commented on src/hotspot/cpu/x86/stubGenerator_x86_64_kyber.cpp:
```diff
@@ -822,10 +855,64 @@ address generate_kyber12To16_avx512(StubGenerator *stubgen,
+    __ lea(perms, ExternalAddress(kyberAvx512_12To16AndAddr()));
+    __ evmovdquq(xmm22, Address(perms), Assembler::AVX_512bit);
+
+    __ BIND(VBMILoop);
```
Just a note on LoopAlignment, there are multiple moving parts here, first aligning starting addresses of loop to 64 ([recommendation from Zen5 optimization guide](https://docs.amd.com/v/u/en-US/58455_1.00) section 2.8.3) ensure small loop bodies are not split-across the cache line, if that happens then there is a code entry penalty since for first iteration of loop front-end will have to read multiple L1I cachelines, once its decoded and uops are part of Op-cache (AMD) or DSB (Intel) then uops stream for successive loop iterations are emitted from op-cache. Since op-cache is shared b/w 2 HW threads in SMT configuration hence in case of noisy neighbor scenarios or context-switches we may hit code-entry penalty during lifetime of loop. 

So its advisable to add alignment in this case for other labels before loops we already have OptoLoopAlignment in place.

jatin-bhateja replied:
> > Better to align loop sarting address to OptoLoopAlignment
> 
> For parity, should I do this for the other labels in the file as well?
> 
> > I will run the micro benchmark on AMD Turin and report back by early next week.
> 
> That would be great, thank you for doing this!

Here are the score on Turin.

```
Baseline:
Benchmark                                    (algorithm)  (keyLength)  (provider)   Mode  Cnt      Score   Error  Units
KeyPairGeneratorBench.MLKEM.generateKeyPair   ML-KEM-512            0              thrpt    2  62235.790          ops/s
KeyPairGeneratorBench.MLKEM.generateKeyPair   ML-KEM-768            0              thrpt    2  38238.390          ops/s
KeyPairGeneratorBench.MLKEM.generateKeyPair  ML-KEM-1024            0              thrpt    2  24725.512          ops/s

Withopt:
Benchmark                                    (algorithm)  (keyLength)  (provider)   Mode  Cnt      Score   Error  Units
KeyPairGeneratorBench.MLKEM.generateKeyPair   ML-KEM-512            0              thrpt    2  62483.697          ops/s
KeyPairGeneratorBench.MLKEM.generateKeyPair   ML-KEM-768            0              thrpt    2  38464.272          ops/s
KeyPairGeneratorBench.MLKEM.generateKeyPair  ML-KEM-1024            0              thrpt    2  24702.044          ops/s
```

System was set at fixed frequency of 2.7 Ghz during benchmarking.

jatin-bhateja replied:
Thanks for the sharp details and analysis.

smemery commented:
/integrate

jatin-bhateja commented:
/sponsor

stooke commented:
/backport jdk25u-dev
